### ISSUE TITLE
$embed doesn't seem to be working right

### ISSUE THREAD

**Issue by ryandesign** [AppleCommander/bastools#24]
Thank you for making these utilities; I'm delighted. The `$embed` functionality is exactly what I was looking for, unfortunately I'm having trouble getting it to work with bt 0.3.1.

```
printf $'\x60' > rts
echo '1 PRINT "BEFORE CALL"' > hello
echo '2 $embed file="rts", moveto="$300"' >> hello
echo '3 PRINT "AFTER CALL"' >> hello
applecommander -dos140 test.dsk
bt hello --applesingle --stdout | applecommander -as test.dsk hello
```

If I run this program:

```
]RUN HELLO
BEFORE CALL

?UNDEF'D STATEMENT ERROR IN 2
```

If I list this program:

```
]LIST

 1  PRINT "BEFORE CALL"
 2  CALL 2110: GOTO 3
 15493  SPEED= DEL = SPEED= ? DEL
     > SPEED= DEL ? SPEED=
 937  DEL C COLOR=
 24830
 17985 TER CALL"
```

**Comment by ryandesign** [comment]
What I expected to see when I ran the program:

```
]RUN HELLO
BEFORE CALL
AFTER CALL
```

What I expected to see when I listed the program:

```
]LIST

 1  PRINT "BEFORE CALL"
 2  CALL 2110: GOTO 3
[various garbage]
 3  PRINT "AFTER CALL"
```

-----

Looking at a simpler program:

```
printf $'\x60' > rts
echo '1 $embed file="rts", moveto="$300"' > hello
echo '2 END' >> hello
applecommander -dos140 test.dsk
bt hello --applesingle --stdout | applecommander -as test.dsk hello
```

The file that is produced is:

```
Offset   Hex Data                                          Characters
=======  ================================================  =================
$000000  2D 08 01 00 8C 32 30 39  31 3A AB 32 00 A9 2B 85  -....209 1:+2.)+.
$000010  3C A9 08 85 3D A9 2C 85  3E A9 08 85 3F A9 00 85  <)..=),. >)..?)..
$000020  42 A9 03 85 43 A0 00 4C  2C FE 60 00 33 08 02 00  B)..C .L ,~`.3...
$000030  80 00 00 00 .. .. .. ..  .. .. .. .. .. .. .. ..  ....             
** END **
```

But when I load the program in an emulator and save the memory to disk...

```
] LOAD HELLO
] BSAVE HELLO.BSAVE,A2049,L52
```

...the memory contents is different:

```
Offset   Hex Data                                          Characters
=======  ================================================  =================
$000000  0E 08 01 00 8C 32 30 39  31 3A AB 32 00 20 08 85  .....209 1:+2. ..
$000010  3C A9 08 85 3D A9 2C 85  3E A9 08 85 3F A9 00 28  <)..=),. >)..?).(
$000020  08 A9 03 85 43 A0 00 31  08 FE 60 00 33 08 02 00  .)..C .1 .~`.3...
$000030  80 00 00 00 .. .. .. ..  .. .. .. .. .. .. .. ..  ....             
** END **
```

Upon loading the file, Applesoft appears to have attempted to recompute the "next line" pointers, which makes sense since the Applesoft start address (memory locations $103–$104) in effect when the file was saved might not be the same as the start address that's in effect now. But since it doesn't know about the binary data bt inserted between the lines, it can't account for those and overwrites that binary data and miscomputes line numbers from there on.

In the hello file as originally produced, the first bytes 2D 08 are the pointer to the next line, which bt had computed to be after the binary data at the end of line 1. After processing by Applesoft, the first bytes have changed to 0E 08, which is the beginning of the binary data and just after the real line 1 data; Applesoft has replaced the first 2 bytes of the binary data (A9 2B) with its computation of the next line pointer (20 08), and so on.

Perhaps a solution is to place all of the binary data at the end of the file?

**Comment by a2geek** [comment]
Sorry for the delay! 

I tried doing the embed at the end of the file: 
```
1 GOSUB 10
2 PRINT "BEFORE CALL"
3 CALL 768
4 PRINT "AFTER CALL"
5 END
10 $embed file="rts", moveto="$300"
```
... and I just figured out that I'm seeing the exact same thing. 

However (don't know if this helps you or not), ProDOS does not seem to do that. DOS 3.3 definitely does. So if ProDOS is possible for whatever you are doing, that would be worth investigating.

There is definitely a bit around the address computation, however. I did have to patch that -- the address that gets called in the `$embed` ends up being the actual code instead of the move routine. I think I have a patch for it and I'll commit that now as a partial fix. 

Uncertain if anything can be done about DOS ... except maybe note that it's an unsolved problem to warn others.
